### PR TITLE
Fix #6677: Critical bug with VM and universes

### LINKED ISSUE #6677: Critical bug with VM and universes
```coq
Require Import Hurkens.

Set Universe Polymorphism.
Set Printing Universes.
Set Printing All.

Definition T@{i} := Type@{i}.
Definition U@{i} := (T@{i} <: Type@{i}).
Definition eqU@{i j} : @eq T@{j} U@{i} T@{i} := eq_refl.
Definition bad : False := TypeNeqSmallType.paradox U eqU.
Print Assumptions bad.
```

We are probably missing a universe level substitution.

It seems at least 8.6, 8.7 and master are impacted. I'm probably responsible for this one. Sorry about that! Although maybe it was already there in the original VM univ code, I haven't checked. While I should have realized I was doing something wrong, I'm not confident that with the current universe API, I won't be doing this kind of mistakes again, unfortunately.

Something that I find very surprising, and that led me to find this bug, is that the kernel accepts terms with unsubstituted universe levels (`Var(i)`). How can that be, if the kernel interface is safe?

```coq
Set Universe Polymorphism.
Set Printing Universes.
Definition U@{i} := Type@{i}.
Definition V := Eval vm_compute in U@{Set}.
Print V.
```

Note that this last example uses the VM bug only in the pretyper, not the kernel. So why isn't `V` rejected ?!

### PR DESCRIPTION
**Pull request by maximedenes**
This bug was present since the first patch adding universe polymorphism
handling in the VM (Coq 8.5). Note that unsoundness can probably be
observed even without universe polymorphism.

### CODE REVIEW

File: kernel/cbytecodes.ml [hunk 1]
```diff
@@ -200,7 +200,7 @@ let rec pp_struct_const = function
   | Const_bn (i,t) ->
      int i ++ surround (prvect_with_sep pr_comma pp_struct_const t)
   | Const_univ_level l -> Univ.Level.pr l
-  | Const_type u -> str "Type@{" ++ Univ.pr_uni u ++ str "}"
+  | Const_type(u,_) -> str "Type@{" ++ Univ.pr_uni u ++ str "}"
```
Comment (SkySkimmer) [inline kernel/cbytecodes.ml]: What is this printer for? Is it OK to forget the int list?

**Review by SkySkimmer — COMMENTED** [review]
Test suite addition?

**Comment by Zimmi48** [comment]
Please have this finished on Monday or we will likely be forced to delay 8.7.2.

**Comment by SkySkimmer** [comment]
Someone who knows the VM should review this.

**Review by mattam82 — APPROVED** [review]
LGTM, as far as I can understand the max_var_pred/compact implementation :)

**Comment by mattam82** [comment]
I see that.
Le lun. 12 févr. 2018 à 14:22, Maxime Dénès <notifications@github.com> a
écrit :

**Comment by maximedenes** [comment]
I already discussed the VM part with @bgregoir. A review of the new code in `kernel/univ.ml` is welcome, however.

**Comment by maximedenes** [comment]
@ppedrot I'm afraid I'm going to give priority to this on #6230. I'll help with the small conflicts, though.

**Comment by mattam82** [comment]
@maximedenes that's what I checked

**Comment by maximedenes** [comment]
Great! Anyway, this patch doesn't touch anything else than the code that is already unsound... :)

**Comment by ppedrot** [comment]
@maximedenes That's fine with me, and I actually support fixing soundness bugs before optimizations! I'll perform a rebase on my side.

**Comment by mattam82** [comment]
As far as I can tell, the other uses of Universe.levels in the codebase are ok.

**Comment by JasonGross** [comment]
> Note that unsoundness can probably be
observed even without universe polymorphism.

Test case without universe polymorphism:
```coq
Require Import Coq.Logic.Hurkens.
Universes i j.
Definition hparadox : forall A : Type@{i}, Type@{i} = A :> Type@{j} -> False
  := Hurkens.TypeNeqSmallType.paradox.
Check hparadox (Type@{i} <: Type@{i}) eq_refl : False.
```

**Comment by Zimmi48** [comment]
It could be worth adding the other test cases of @JasonGross without universe polymorphism to the test-suite. WDYT @maximedenes?

**Comment by maximedenes** [comment]
> It could be worth adding the other test cases of @JasonGross without universe polymorphism to the test-suite. WDYT @maximedenes?

No real opinion, it hits exactly the same code path.

**Pull request merged**